### PR TITLE
ncm-systemd: Add support for Socket in units

### PR DESCRIPTION
**Pull request by wpoely86**
This add supports for a `[Socket]` section in a systemd unit.

It builds on top of and requires #1396 

Please have look @stdweird

### CODE REVIEW

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (stdweird) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:260]: there is an absolute_file_path type somewhre

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (stdweird) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:265]: use `choice`

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (stdweird) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:271]: these probably can also be limited `(0..1023)` (ot sure anyone wants to set the sticky bit, otherwise stop at 511

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (stdweird) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:302]: choice

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (stdweird) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:316]: `service` -> `Service`, and `Accept=no` -> `Accept=false`

**Comment by stdweird** [comment]
@wpoely86 minor remarks, thanks for this!

**Comment by wpoely86** [comment]
@stdweird I just noticed a thing: `ExecStartPost` (and friends). I've defined them as a list as systemd allows multiple lines of those. In the `Service` section, they are also defined but as string. The tt files do however have them in `list_of_lines` (together with `ExecStart` which is not correct according to the man page). So it's unclear if that can work at all to me.
I can change it, but that's a breaking change.

**Comment by stdweird** [comment]
@wpoely86 `ExecStart` and friens should work as list_of_lines. it's doucmented and even has an example.

**Comment by wpoely86** [comment]
@stdweird can you point tot that documentation? I can't find it.

As far as I understand how this works (and my knowledge is pretty limited), it shouldn't work.

Multiple `ExecStart` are in any case forbidden by systemd (according to the man page).

**Comment by stdweird** [comment]
@wpoely86 where is that stated?
```
If more than one command is specified, the commands are invoked sequentially in the order they appear in the unit file. If one of the commands fails (and is not prefixed with "-"), other lines are not executed, and the unit is considered failed.
```
and a bit lower
```
Example:

Environment=ONE='one' "TWO='two two' too" THREE=
ExecStart=/bin/echo ${ONE} ${TWO} ${THREE}
ExecStart=/bin/echo $ONE $TWO $THREE

This results in /bin/echo being called twice, the first time with arguments "'one'", "'two two' too", "", and the second time with arguments "one", "two two", "too". 
```
i don't know if i ever tried this

**Comment by wpoely86** [comment]
```
       ExecStartPre=, ExecStartPost=
           Additional commands that are executed before or after the command in ExecStart=, respectively. Syntax is the same as for ExecStart=, except that multiple command lines are allowed and the commands are executed one after the other, serially.
```

**Comment by wpoely86** [comment]
OK, the docs are confusing:
```
Unless Type= is oneshot, exactly one command must be given. When Type=oneshot is used, zero or more commands may be specified. 
```

So it's fine, but only when it's a oneshot.

But in any case, how you specify this? Currently `ExecStart` is a string?

**Comment by stdweird** [comment]
@wpoely86 ok, i'm also confused now what should be working. 
but indeed, either the schema becomes a list of strings or the TT shouldn't use list_of_lines. for now, i would remove ExecStart from list_of_lines and see what the unittests do.

**Comment by wpoely86** [comment]
The unit tests keep on working just fine if you remove it. I'm pretty sure it doesn't work right now. You can only specify a single command. The `? string` needs to be `? string[]`.

I will open a PR for it once this is merged.

**Comment by stdweird** [comment]
@wpoely86 thanks. can you rebase on the other PR also cleanup the commit history here.

**Comment by wpoely86** [comment]
@stdweird done

**Comment by stdweird** [comment]
@wpoely86 for me this is ok. i'll let @ned21 decide if using long type and assumig users will know what they are doing is the way to go. i'm in favour, but i understand that it's error prone

**Comment by stdweird** [comment]
@wpoely86 @ned21 actuallu, this is already in use elsewhere https://github.com/quattor/configuration-modules-core/blob/master/ncm-metaconfig/src/main/pan/components/metaconfig/schema.pan#L91
so i'm going to approve this PR

**Comment by wpoely86** [comment]
@stdweird all checks finally passed!

**Comment by stdweird** [comment]
@wpoely86 clearly the unittests do not try to include the component config.pan
the `%{}` mahgic in that file binds to `systemd_component` type, the curren scheam has `component_systemd` type. can you fix the schema ?

**Comment by wpoely86** [comment]
@stdweird fixed and tests are go

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (ned21) [inline ncm-systemd/src/main/pan/components/systemd/schema.pan:152]: Looks like systemd only accepts relative paths here, and `..` is banned - should we have a type that enforces that?

**Review by ned21 — COMMENTED** [review]
Can you squash e769272 into the previous commit please since it seems to be a fixup?

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (wpoely86) [reply ncm-systemd/src/main/pan/components/systemd/schema.pan:152]: I can add something like:

```
# type for a relative directory (no leading /) and may not include ".."
type ${project.artifactId}_relative_directory = string with !match(SELF, '(^/|\.\.)');
```

But currently we don't really check anything like this. The `WorkingDirectory` and `RootDirectory` are also just strings.

**Comment by wpoely86** [comment]
@ned21 commits squashed

File: ncm-systemd/src/main/pan/components/systemd/schema.pan
Comment (ned21) [reply ncm-systemd/src/main/pan/components/systemd/schema.pan:152]: Yep,  I think that would make a lot of sense if you can add it easily enough.

Comment (wpoely86) [reply ncm-systemd/src/main/pan/components/systemd/schema.pan]: Done. Please have a look.

**Review by ned21 — DISMISSED** [review]

**Comment by kwaegema** [comment]
@wpoely86 has merge conflicts now

**Comment by wpoely86** [comment]
@kwaegema fixed.

**Comment by ned21** [comment]
Any particular reason you merged from master instead of rebaseing ? It keeps the commit history easier to read if it's linear?

**Comment by wpoely86** [comment]
@ned21 rebased instead of merged now.

**Review by ned21 — APPROVED** [review]

**Pull request merged**
